### PR TITLE
Modifications to the ensemble pipeline

### PR DESCRIPTION
**Pull request by lizgzil**
- Add new 21.04.01 ensemble config
- Use wellcomeml votingclassifier in ensemble model
- Allow option for using a prediction probability threshold
- Some fixes to GrantTagger including converting predictions to integers (previously floats)

I noticed a bug in GrantTagger in which we scaled the vectors for the bert/scibert+naive bayes case, but then didn't include this in `transform`. Since the analysis after this uses the results saved during the training, these are still accurate, but when I tried to replicate them I couldn't. Thus `GrantTagger` has been modified to save out the trained scalers for these cases, and use them if loaded.

After this, I replicated the results from the ensemble parameter exploration notebook, namely that the ensemble model with config ` 21.04.01` gives:
```
             precision    recall  f1-score   support

           0       0.89      0.91      0.90        77
           1       0.91      0.89      0.90        80

    accuracy                           0.90       157
   macro avg       0.90      0.90      0.90       157
weighted avg       0.90      0.90      0.90       157

[[70  7]
 [ 9 71]]
 ```

### CODE REVIEW

**Comment by lizgzil** [comment]
[Tech model refactor - ensemble of models](https://trello.com/c/TRUJH59s/1200-tech-model-refactor-ensemble-of-models)

File: nutrition_labels/ensemble_grant_tagger.py [hunk 1]
```diff
@@ -122,25 +103,32 @@ def output_tagged_grants(self, output_path):
...
     config = configparser.ConfigParser()
     config.read(args.config_path)
 
-    datestamp = datetime.now().date().strftime('%y%m%d')
+    config_version = "".join(config["DEFAULT"]["version"].split("."))[2:]
 
     grants_data_path = config["prediction_data"]["grants_data_path"]
     input_file_name = os.path.basename(grants_data_path).split('.')[0]
-    output_path = f'data/processed/ensemble/{datestamp}/{input_file_name}_tagged.csv'
+    output_path = f'data/processed/ensemble/{config_version}/{input_file_name}_tagged.csv'
```
Comment (nsorros) [inline nutrition_labels/ensemble_grant_tagger.py:118]: interesting, do you now version your data like that? this is a good step towards reproducibility

File: nutrition_labels/grant_tagger.py [hunk 2]
```diff
@@ -172,9 +180,12 @@ def fit_transform(self, data, train_data_id="Internal ID"):
             print("Vectorizer type not recognised")
         self.X_vect = self.vectorizer.fit_transform(self.X)
 
+        # For BERT/SciBERT + naive bayes the values need to be between 0 and 1, 
+        # so scale the values.
+        self.scaler = None
         if "bert" in self.vectorizer_type and self.classifier_type == "naive_bayes":
-            scaler = MinMaxScaler()
-            self.X_vect = scaler.fit_transform(self.X_vect)
+            self.scaler = MinMaxScaler()
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:187]: you can add that to be part of the vectorizer if you define the vectorizer as a pipeline. pipelines in sklearn are composable i.e. one pipeline can contain multiple pipelines and your vectorizer can be a pipeline

File: tests/test_ensemble_grant_tagger.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                {'id':3, 'text_1': 'cat shoe', 'text_2': '', 'label': 0},
+                {'id':4, 'text_1': 'dog dog skate cat shoe sandal', 'text_2': '', 'label': 0},
+                {'id':5, 'text_1': 'sandal', 'text_2': '', 'label': 0},
+                {'id':6, 'text_1': None, 'text_2': None, 'label': 0}
+            ]
+        )
+
+def test_load_grants_text():
+
+    with tempfile.TemporaryDirectory() as tmp_dir:
```
Comment (nsorros) [inline tests/test_ensemble_grant_tagger.py:25]: I think pytest has a tmp_path fixture that is provided automatically so if you do `test_load_grants_text(tmp_path)` you can remove this line

see https://docs.pytest.org/en/stable/tmpdir.html

File: nutrition_labels/grant_tagger.py [hunk 4]
```diff
@@ -319,6 +341,12 @@ def load_model(self, output_path):
             self.model = pickle.load(f)
         with open(os.path.join(output_path, "vectorizer.pickle"), "rb") as f:
             self.vectorizer = pickle.load(f)
+        try:
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:344]: apparently try/except for opening a file instead of checking if it exists is more pythonic https://blog.finxter.com/check-if-a-file-exists-in-python/

File: nutrition_labels/grant_tagger.py [hunk 5]
```diff
@@ -230,7 +241,15 @@ def fit(self, X, y):
         self.model = model.fit(X, y)
 
     def predict(self, X):
-        return self.model.predict(X)
+        y_predict = self.model.predict(X).astype(int)
+        if self.pred_prob_threshold:
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:245]: maybe you can name that just `threshold
`

File: nutrition_labels/grant_tagger.py [hunk 6]
```diff
@@ -230,7 +241,15 @@ def fit(self, X, y):
         self.model = model.fit(X, y)
 
     def predict(self, X):
-        return self.model.predict(X)
+        y_predict = self.model.predict(X).astype(int)
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:244]: this is not needed if the user has set the threshold so I would add it as an else statement after
```
if self.pred_prob_threshold:
    # If the prediction probability is over a threshold then allow a 1
    # prediction to stay as 1, otherwise switch to 0.
    # A prediction of 0 will stay at 0 regardless of probability.
    pred_probs = self.model.predict_proba(X)
    y_predict = y_predict*(np.max(pred_probs, axis=1) >= self.pred_prob_threshold
else:
    y_predict = self.model.predict(X).astype(int)
```

**Review by nsorros — APPROVED** [review]
Nice versioning of data and models. Next step DVC.

Note that you can add the scaler inside the pipeline that defines the vectorizer and pickle them together

File: nutrition_labels/grant_tagger.py (hunk 2 above)
Comment (lizgzil) [reply nutrition_labels/grant_tagger.py:187]: good idea - will add this when I correct https://github.com/wellcometrust/nutrition-labels/issues/72

File: tests/test_ensemble_grant_tagger.py (hunk 3 above)
Comment (lizgzil) [reply tests/test_ensemble_grant_tagger.py:25]: whoa 🤯

File: nutrition_labels/grant_tagger.py (hunk 4 above)
Comment (lizgzil) [reply nutrition_labels/grant_tagger.py:344]: good to know - I hadn't given it much thought. Worrying to think something could happen to my files between checking they exist and opening though!
